Node/core/lib/dialogs/EntityRecognizer.js

```diff
@@ -176,13 +176,35 @@ var EntityRecognizer = (function () {
                 score = Math.min(0.5 + (value.length / utterance.length), 0.9);
             }
             else {
-                var matched = '';
+                var matched = {};
                 tokens.forEach(function (token) {
                     if (value.indexOf(token) >= 0) {
-                        matched += token;
+                        if (!matched[token]) {
+                            matched[token] = 1;
+                        }
                     }
                 });
-                score = matched.length / value.length;
+                var tokenizedValue = value.split(' ');
+                var tokenScore = 0;
+                for (var token in matched) {
+                    tokenizedValue.forEach(function (val) {
+                        if (val.indexOf(token) >= 0 && token.length <= val.length / 2) {
+                            matched[token]--;
+                        }
+                        else if (val.indexOf(token) == -1) {
+                        }
+                        else {
+                            matched[token]++;
+                        }
+                    });
+                }
+                for (var token in matched) {
+                    if (matched[token] > 0) {
+                        tokenScore += token.length;
+                    }
+                }
+                score = tokenScore / value.length;
+                score = score > 1 ? 1 : score;
```

Comment:
For [line 189][L189] through [line 200][L200] we evaluate if the `token` from `matched` exists in and is less than half of the current token (`val`) from the earlier variable `value`. If these are both true, we're assuming it's not relevant to the current value, and we decrement `matched[token]`. If the `token` from `matched` doesn't exist in the current `val`, do nothing. Passing these two cases, the [current `token`][L197] must be larger than half the current `val` and therefore is relevant to the current `val` (see [Stemming and lemmatization][Stanford] from the Stanford NLP Group).

On [line 201][L201] through [line 205][L205] we evaluate which `tokens` are still relevant to the current `value` (relevant meaning they have a score greater than 0). If relevant we'll add the length of the token to `tokenScore`. We then calculate the score in the same way as the previous method (dividing the length of all of the matching tokens by the length of the current `value`). 

As an added precaution, there's a ternary operator for `score` on [line 207][L207] that evaluates if the `score > 1` in which case it changes score to 1.

  [L189]: https://github.com/Microsoft/BotBuilder/pull/3702/files#diff-a5cbf2306473b89ce8f6eb849e7f09c5R189
  [L197]: https://github.com/Microsoft/BotBuilder/pull/3702/files#diff-a5cbf2306473b89ce8f6eb849e7f09c5R197
  [L200]: https://github.com/Microsoft/BotBuilder/pull/3702/files#diff-a5cbf2306473b89ce8f6eb849e7f09c5R200
  [Stanford]: https://nlp.stanford.edu/IR-book/html/htmledition/stemming-and-lemmatization-1.html
  [L201]: https://github.com/Microsoft/BotBuilder/pull/3702/files#diff-a5cbf2306473b89ce8f6eb849e7f09c5R201
  [L205]: https://github.com/Microsoft/BotBuilder/pull/3702/files#diff-a5cbf2306473b89ce8f6eb849e7f09c5R205
  [L207]: https://github.com/Microsoft/BotBuilder/pull/3702/files#diff-a5cbf2306473b89ce8f6eb849e7f09c5R207